File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function checkDist(x1, y1, x2, y2) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}

function solution(numbers, hand) {
    var point = [[3, 1], [0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1], [2, 2], [3, 0], [3, 2]];
    
    var answer = '';
    var left = 10, right = 11;
    
    numbers.forEach(function(item, index) {
        if(item === 1 || item === 4 || item === 7) {
            answer += "L";
            left = item;
        } else if(item === 3 || item === 6 || item === 9) {
            answer += "R";
            right = item;
        } else {
            var cal = checkDist(point[item][0], point[item][1], point[left][0], point[left][1]) - checkDist(point[item][0], point[item][1], point[right][0], point[right][1]);
            
            if(cal === 0) {
                if(hand == "right") {
                    answer += "R";
                    right = item;
                } else {
                    answer += "L";
                    left = item;
                }
            } else if(cal > 0) {
                answer += "R";
                right = item;
            } else {
                answer += "L";
                left = item;
            }
        }
    })
    
    return answer;
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    int[][] point = {{0, 4, 3, 4, 3, 2, 3, 2, 1, 2, 1, 1},
                     {4, 0, 1, 2, 1, 2, 3, 2, 3, 4, 3, 5},
                     {3, 1, 0, 1, 2, 1, 2, 3, 2, 3, 4, 4},
                     {4, 2, 1, 0, 3, 2, 1, 4, 3, 2, 5, 3},
                     {3, 1, 2, 3, 0, 1, 2, 1, 2, 3, 2, 4},
                     {2, 2, 1, 2, 1, 0, 1, 2, 1, 2, 3, 3},
                     {3, 3, 2, 1, 2, 1, 0, 3, 2, 1, 4, 2},
                     {2, 2, 3, 4, 1, 2, 3, 0, 1, 2, 1, 3},
                     {1, 3, 2, 3, 2, 1, 2, 1, 0, 1, 2, 2},
                     {2, 4, 3, 2, 3, 2, 1, 2, 1, 0, 3, 1},
                     {1, 3, 4, 5, 2, 3, 4, 1, 2, 3, 0, 2},
                     {1, 5, 4, 3, 4, 3, 2, 3, 2, 1, 2, 0}};
    
    public String solution(int[] numbers, String hand) {
        int leftNow = 10, rightNow = 11;
        StringBuilder answer = new StringBuilder("");
        
        for(int num : numbers) {
            if(num == 1 || num == 4 || num == 7) {
                answer.append("L");
                leftNow = num;
                continue;
            }
            
            if(num == 3 || num == 6 || num == 9) {
                answer.append("R");
                rightNow = num;
                continue;
            }
            
            int cal = point[num][leftNow] - point[num][rightNow];
            
            if(cal != 0) {
                answer.append(cal > 0 ? "R" : "L");
                
                if(cal > 0) {
                    rightNow = num;
                } else {
                    leftNow = num;
                }
                
                continue;
            }
            
            if(hand.equals("right")) {
                answer.append("R");
                rightNow = num;
            } else {
                answer.append("L");
                leftNow = num;
            }
        }
        
        return answer.toString();
    }
}